### DISCUSSION TITLE
Installation

### DISCUSSION THREAD

**Discussion by jk262626** [nakib/elphbolt#120] [General]
Dear Nakib,
Thank you so much for your effort on a wonderful software, elphbolt!

I'm trying to install elphbolt in HPC system (Linux-rocky8-zen2) and this is the shared cluster system like Expanse. 
Both the spglib and opencoarrays are installed on my HPC system by myself.
And then, I tried to do the next step '3. build' in your instruction with the following procedure. 

1. Load opencoarrays and activate fpm.
2. I have all of the files and folders of your code in my working directory. 
3. Type the command instead of executing fpm_config_caf.sh because I'm not sure if the command '. fpm_config_caf.sh' would work for this shared cluster system.
export FPM_FC=caf
export FPM_FFLAGS="-g -O2 -fPIC -fbounds-check -Wunused -Wconversion -Wunderflow -Wdo-subscript"
4. Type 'fpm build' and then I encountered error like the figure below. The permission for git@github.com was denied. 
![Screenshot 2023-11-08 112204](https://github.com/nakib/elphbolt/assets/138066037/13bf8e50-2550-47f3-8a20-10f9d2926562)
5. So I deleted the following dependencies line in the fpm.toml file, and then the error disappeared and new error came out like the figure below. 
[dependencies]
testify = { git = "git@github.com:nakib/testify.git" } . 
![image](https://github.com/nakib/elphbolt/assets/138066037/5e0b0014-6a66-40be-b41a-b5b25f7c0afd)

Can I get an advice on this? I want to stick to fpm to build elphbolt. 
Thank you!

**Comment by nakib** [comment]
Dear jk262626,

Thanks for your kind words!

As for your questions:

3. `. fpm_config_caf.sh` or, equivalently, `source fpm_config_caf.sh` should work since it just defines a couple of environment variables for the current open shell. I am surprised that it didn't work for you. But manually exporting the flags also work. Additionally, can you confirm that `echo $FPM_FC` and `echo $FPM_FLAGS` are reproducing the contents of `fpm_config_caf.sh`? Also, could you confirm that `which caf` and `which cafrun` return something meaningful?

4. Hmm...this seems strange. Is it happening because your are within a `conda` environment? But let's revisit this since this part is about building the tests.

5. Have you included the header file and linked the shared library of `spglib` as explained here: https://spglib.readthedocs.io/en/latest/install.html ?

You can do the above once and for all by adding the following to your `~/.bashrc` file:

```
export PATH="<path to spglib build directory>:$PATH"
export LD_LIBRARY_PATH="<path to spglib build directory>:$LD_LIBRARY_PATH"
```

Don't forget to `source ~/.bashrc` after the change.

Best,
Nakib

**Comment by jk262626** [comment]
I confirmed what you said. 
![image](https://github.com/nakib/elphbolt/assets/138066037/32e8e319-8688-4b31-b99f-e565967274dd)

I got the source code by 
$ git clone https://github.com/spglib/spglib.git
and then tried to compile but encountered the error like this.
![image](https://github.com/nakib/elphbolt/assets/138066037/9bebf044-bf6a-47e3-9285-0d6673485e6b)

Should I need to do this? I skip this. 
% tar xvfz spglib-2.0.0.tar.gz
% cd spglib-2.0.0

**Reply by nakib** [reply]
Ok, thanks for checking.

Regarding `spglib`, it looks like the build system is trying to update something but failing to connect to github. Could be a related issue to the error you got when `fpm` tried to clone the `testify` repo.

You mentioned that `spglib` is already available on your HPC system as a module. Could you check where `libsymspg.so` is located? That is, what is the output of `locate libsymspg.so` ?

**Reply by jk262626** [reply]
I checked!
![image](https://github.com/nakib/elphbolt/assets/138066037/10d918b5-c530-4fa2-a6c5-9ff1677aab65)

I guess I already have libsymspg.so (FYI, I've installed it by using $ conda install -c conda-forge spglib).
If I correctly have it on my system, then should I add the following commands to the end of the ~/.bashrc file and then type source ~/.bashrc?
export PATH="path to spglib build directory:$PATH"
export LD_LIBRARY_PATH="path to spglib build directory:$LD_LIBRARY_PATH"

That "path to spglib build directory" would be one of the directories in the picture?

**Reply by nakib** [reply]
Since I am not a `conda` user, I am actually not sure if it sets up the environments such that the shared library (`libsymspg.so`) will be findable in your `fpm` environment. From the 2nd line in the screenshot, it looks like it is available.  You can double check if `echo $LD_LIBRARY_PATH` includes this location.

It looks like you have managed to install `spglib` successfully. I think you can retry building `elphbolt`.

**Comment by nakib** [comment]
Could you please try to build the latest develop branch? I made 2 changes to the `fpm.toml` file:

1. I added the following in `fpm.toml`. You are using a new version of `fpm` which by default does not allow compilation of code with an implicit interface. This is why the file `spglib_wrapper.f90` was not being built. With the following, this error should go away:
```
...

[fortran]
implicit-external = true #Needed for v>=0.7 fpm

...
```
2. With the following change, `fpm` will build `testify` from a local copy instead of trying to pull from the remote repository.
```
...

[dependencies]
#Pull from github
#testify = { git = "git@github.com:nakib/testify.git" }
#Use local copy from thirdparty directory
testify = {path = "thirdparty/testify"}

...
```

Would you please let me know if this solves you issues?

Best,
Nakib

**Comment by jk262626** [comment]
Thank you so much for the update!
![image](https://github.com/nakib/elphbolt/assets/138066037/ce48c2cc-5d00-46a9-9026-9f220e170622)
Following your instruction, f90 files can be compiled but I encountered the error at the end like this.

**Reply by nakib** [reply]
This is because you don't have libsymspg, libblas, and liblapack available. Have you loaded the necessary modules? Typically openblas is available as a module. You also mentioned spglib is available as a module. Please try loading these first.

**Reply by jk262626** [reply]
Here is the directory of each libraries.
![image](https://github.com/nakib/elphbolt/assets/138066037/47a31100-327c-47ff-b73e-a211ae933dc7)

They are all included in /home/jkim3/miniconda3/envs/fpm/lib, so I added the path to the ~/.bashrc file and typed source ~/.bashrc.
export PATH="/home/jkim3/miniconda3/envs/fpm/lib:$PATH"
export LD_LIBRARY_PATH="/home/jkim3/miniconda3/envs/fpm/lib:$LD_LIBRARY_PATH"
![image](https://github.com/nakib/elphbolt/assets/138066037/9876c9c7-a0c8-416f-a885-dc56a323883f)

And then I tried to build again, but have same error.

**Reply by nakib** [reply]
Hmm...this is quite surprising.

Can you do a `module load spglib`? I'm assuming that's what the module is called on your HPC stack. The thing is that spglib requires libblas and liblapack.

**Reply by jk262626** [reply]
Yea, I also loaded openblas module for both libblas and liblapack but even the errors for those two still appear.
I should check this with my cluster manager and then I will reach out to you again!

**Comment by jk262626** [comment]
Hi Nakib,
I have all of libsymspg, libblas, and liblapack in my conda environment and also have openblas, but I still have the same error like the figure below.
![image](https://github.com/nakib/elphbolt/assets/138066037/d0d1f4af-c449-4cc7-9c63-07fccb6d75f8)
I already have changed the PATH and LD_LIBRARY_PATH to where library files are located like the below, but it is still finding the library files in /usr/bin/ld. 
export PATH="/home/jkim3/miniconda3/envs/fpm/lib:$PATH"
export LD_LIBRARY_PATH="/home/jkim3/miniconda3/envs/fpm/lib:$LD_LIBRARY_PATH"

**Reply by jk262626** [reply]
What is the "path to spglib build directory"?

**Reply by jk262626** [reply]
I have also tried to build elphbolt by using cmake, but had same error. 
![image](https://github.com/nakib/elphbolt/assets/138066037/d027d859-9223-47f6-a090-83efdc09b897)

**Reply by nakib** [reply]
The issue seems to be related to the compiler not finding the libraries to link to. Let's continue with your cmake build. Can you try the following replacement in your CMakeLists.txt file (I'm assuming your have openblas; change to whatever flavor you have):

```
...

#Set flags for caf
if(CMAKE_Fortran_COMPILER MATCHES "caf")
  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -g -O2 -cpp -fPIC -fbounds-check -Wunused -Wconversion -Wunderflow -Wdo-subscript -L<enter path to the openblas lib directory> -lopenblas")
  #set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -g -O2 -fPIC -fbounds-check -Wunused -Wconversion -Wunderflow -Wdo-subscript")
endif()

#Set spg library
find_library(SPG_LIB NAMES symspg REQUIRED)
#find_library(BLAS_LIB NAMES blas REQUIRED)
#find_library(LAPACK_LIB NAMES lapack REQUIRED)

#Set library links
#set(LINKS "-lblas -llapack ${SPG_LIB}")
set(LINKS "${SPG_LIB}")

...
```

Basically, above I am explicitly setting the linker instead of asking cmake to find the libraries.

Could you also reach out to your cluster manager if the above does not work? But also let me know how it goes.

As for your other question, "path to spglib build directory" refers to where the `libsymspg.a`,  `libsymspg.so`, etc. files are.

Best regards,
Nakib

**Comment by jk262626** [comment]
Hi Nakib, 
Thanks to your comments, it seems to be linked to the library now. 
I modified Cmake file like this. 
```
#Set flags for caf
if(CMAKE_Fortran_COMPILER MATCHES "caf")
  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -g -O2 -cpp -fPIC -fbounds-check -Wunused -Wconversion -Wunderflow -Wdo-subscript -L/home/jkim3/miniconda3/pkgs/libopenblas-0.3.21-h043d6bf_0/lib/-lopenblas")
  #set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -g -O2 -fPIC -fbounds-check -Wunused -Wconversion -Wunderflow -Wdo-subscript")
endif()

#Set spg library
find_library(SPG_LIB NAMES symspg REQUIRED)
#find_library(BLAS_LIB NAMES blas REQUIRED)
#find_library(LAPACK_LIB NAMES lapack REQUIRED)

#Set library links
#set(LINKS "-lblas -llapack ${SPG_LIB}")
set(LINKS "${SPG_LIB}")
``` 

Now, I'm getting a new error like this figure below!
![image](https://github.com/nakib/elphbolt/assets/138066037/82f24c1f-ab94-4c35-9b99-f169ef195525)

**Comment by nakib** [comment]
Is there a space missing before -lopenblas in the linking command?

**Reply by jk262626** [reply]
Yes I was missing some space

**Comment by jk262626** [comment]
Awesome!! I finally can compile it.
![image](https://github.com/nakib/elphbolt/assets/138066037/c0842662-0097-4041-b326-a6ed960672b4)
![image](https://github.com/nakib/elphbolt/assets/138066037/59e214fb-55bf-46db-afbd-4f315a44ef4c)
I'm running ctest now.

**Comment by nakib** [comment]
Great!

I'll let you know when I figure out why the fpm build didn't work for you.

**Reply by jk262626** [reply]
Nice thanks!

**Comment by jk262626** [comment]
![image](https://github.com/nakib/elphbolt/assets/138066037/676d6d8f-6f79-4978-bb91-d12b6f5aa8e8)
I encountered some errors after running ctest. 
Should I address this or can I just move forward with Si example?

**Comment by nakib** [comment]
The tests should pass. Could you share the output of the test? I can take a
look on Monday.

**Reply by jk262626** [reply]
[CTestCostData.txt](https://github.com/nakib/elphbolt/files/13531850/CTestCostData.txt)
[LastTest.log](https://github.com/nakib/elphbolt/files/13531851/LastTest.log)
[LastTestsFailed.log](https://github.com/nakib/elphbolt/files/13531852/LastTestsFailed.log)
Thanks!

**Reply by nakib** [reply]
I was not able to reproduce the failing of `test_misc` on any of my linux machines. As you can see for yourself, the same tests also pass through the github actions. Suprisingly, the BTE regression test passed in your case, the output of which you can see in the `LastTest.log` file.

Could you try rebuilding following a `make clean` inside the build directory?

**Reply by nakib** [reply]
If possible, please send the output of `ctest --rerun-failed --output-on-failure`.

Best,
Nakib

**Reply by nakib** [reply]
Ok, after some head scratching and searching, I think your issue is the one described here: https://github.com/sourceryinstitute/OpenCoarrays/issues/778

The issue only affects single image calculations which expains why the single image unit tests failed to start but the multi-image regression test passed.

In case you want to fix this, while I am not the best person to fix issues related to OpenCoarrays, I can suggest that you build it from source using mpich as the backend instead of openmpi. You can use the following to build from source (taken from elphbolt's github actions script/Dockerfile):
```
git clone https://github.com/sourceryinstitute/OpenCoarrays && \
mkdir OpenCoarrays/opencoarrays-install  && \
cd OpenCoarrays/opencoarrays-install && \
git checkout tags/2.10.1 && \
FC="$(command -v gfortran)" CC="$(command -v gcc)" cmake .. && \
make install && \
caf --version && \
cafrun --version
```

See if this helps.

Best,
Nakib

**Reply by jk262626** [reply]
Thank you so much! 
Can I just move forward if I don't need to use single image calculation?
What do single image calculation and multi image regression mean?

**Reply by jk262626** [reply]
[LastTest.log](https://github.com/nakib/elphbolt/files/13550566/LastTest.log)
[LastTestsFailed.log](https://github.com/nakib/elphbolt/files/13550567/LastTestsFailed.log)
[CTestCostData.txt](https://github.com/nakib/elphbolt/files/13550568/CTestCostData.txt)
![image](https://github.com/nakib/elphbolt/assets/138066037/cb908617-3034-4de5-82df-66b42b3480fc)

Here is after ctest --rerun-failed --output-on-failure

**Reply by nakib** [reply]
Thanks. For some reason, this is not showing the full error message. But going by the symptoms, I think the issue is related to a bug in the OpenCoarrays library (https://github.com/sourceryinstitute/OpenCoarrays/issues/778) when using the OpenMPI backend.

An image in coarrays parlance is a single unit of computing. See, for example, here: https://docs.nersc.gov/development/programming-models/coarrays/. Roughly, you can think of an image as an MPI processes, if you are familiar with the SPMD parallelization model. I say rougly since coarrays is an implementation of the PGAS model (https://en.wikipedia.org/wiki/Partitioned_global_address_space), not SPMD.

The unit tests in elphbolt test the general purpose procedures. They are all serial procedures and, as such, the tests are run on a single image. They finish in less than a second. In your case, 3764.72 seconds pass, and the unit tests never starts running. The regression test is a full calculation of the electron-phonon BTEs, start to finish. When you run this test, it runs on 2 images, dividing some of the tasks among them. For production level BTE calculations, you will need dozen to hundreds of images, depending on the size of your crystal unit cell. So I think you can proceed with your current build for your research work. That said, in the ideal case, the issue should be fixed. You can try a different version of the OpenCoarrays library to see if that fixes the issue.

Best regards,
Nakib

**Reply by jk262626** [reply]
I understand now. Thank you for the clarification.
I tried running AlN calculation with full set of inputs and the calculation was terminated incorrectly. 
Here is some part of the slurm file. Please take a look at it. 

Number of coarray images =    20
___________________________________________________________________________
______________________________________________________Setting up crystal...
At line 200 of file /home/jkim3/elphbolt/src/crystal.f90 (unit = 1, file = 'input.nml')
Fortran runtime error: Cannot match namelist object name autoisotopes

Error termination. Backtrace:
At line 200 of file /home/jkim3/elphbolt/src/crystal.f90 (unit = 1, file = 'input.nml')
Fortran runtime error: Cannot match namelist object name autoisotopes

Error termination. Backtrace:
At line 200 of file /home/jkim3/elphbolt/src/crystal.f90 (unit = 1, file = 'input.nml')
Fortran runtime error: Cannot match namelist object name autoisotopes

**Reply by nakib** [reply]
Since v1.1, the flag `autoisotopes` does not exist. Instead, set the flag `VCA = .true.` to tell the code to use the isotopic average of each basis atom. Alternatively, set the flag `DIB = .true.` to tell the code to use the dominant isotopic mass.

Best,
Nakib

**Reply by jk262626** [reply]
Thanks you so much! I resolved by using DIB = .true.
..............
___________________________________________________________________________
______________________________________________________Setting up Wannier...
Wannier data from EPW will be read.
At line 312 of file /home/jkim3/elphbolt/src/wannier.f90 (unit = 1, file = 'epwdata.fmt')
Fortran runtime error: Bad real number in item 1 of list input

Error termination. Backtrace:
At line 312 of file /home/jkim3/elphbolt/src/wannier.f90 (unit = 1, file = 'epwdata.fmt')
Fortran runtime error: Bad real number in item 1 of list input

Error termination. Backtrace:
At line 312 of file /home/jkim3/elphbolt/src/wannier.f90 (unit = 1, file = 'epwdata.fmt')
Fortran runtime error: Bad real number in item 1 of list input

I encountered error like this for this time

**Reply by nakib** [reply]
What is the content of the file `epwdata.fmt`? Can you show the first few lines here?

**Reply by jk262626** [reply]
Oh I missed data from the epw dft calculation. I will let you know after doing that again. Thanks!
